### ISSUE TITLE
Reduce synchronisation overhead in MemValueFactory

### ISSUE THREAD

**Issue by ansell** [eclipse/rdf4j#735]
The full synchronisation in MemValueFactory may be unnecessarily reducing concurrency. There are a range of non-blocking concurrency primitives available in the JDK that were not available in its first iteration a number of years ago so it may benefit from a redesign.

**Comment by boyarintsev** [comment]
Hi guys, are there any plans to merge @ansell 's commit into master / release? This could be very helpful. I've always got 30 threads waiting on one of these locks (getURI, etc.)

**Comment by jeenbroekstra** [comment]
@ansell are you available to look at this and possibly submit a PR? I guess it will need to be re-done, given that this is from before the repository restructure, and the `MemValueFactory` has moved to the rdf4j-storage repo.

**Comment by ansell** [comment]
I will put this on my todo list. The commit referenced above is not production quality, just exploring what would happen if some things changed.

**Comment by boyarintsev** [comment]
Hi guys! @jeenbroekstra thanks fro keeping an eye on this. I was very glad to see the issue in the earliest to do list, but when I followed the link, I wasn't able to find the ticket. Was it deferred? Thanks!

**Comment by ansell** [comment]
Hi @boyarintsev , The draft commit ID referenced above is as far as I was able to get:

https://github.com/ansell/rdf4j/commit/ffc3b5f40fa4ba92d4571c3582fecab7e37ae250 

It isn't much, but could be put up for a Pull Request and reviewed if necessary.

There are many more things that could be done in terms of synchronisation bottlenecks, but I do not foresee me having much time to work on them. Some of the consistency in the currently approach seems like it may derive from being globally synchronised at various points. Removing them did not turn out to be an easy process for me and I had to move on to other issues.
